iothub/device/src/Transport/Amqp/AmqpUnit.cs

```diff
@@ -245,7 +251,7 @@ private async Task EnsureMessageReceivingLinkIsOpenAsync(CancellationToken cance
             if (Logging.IsEnabled)
                 Logging.Enter(this, nameof(EnsureMessageReceivingLinkIsOpenAsync));
 
-            AmqpIotSession amqpIotSession = await EnsureSessionIsOpenAsync(cancellationToken).ConfigureAwait(false);
+            _amqpIotSession = await EnsureSessionIsOpenAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Instead of saving the reference on a local variable, use the class variable to perform all opening and closing operations. The session object is opened under a lock and the transport layer disconnection and retry is also handled under locks.